module3/homework/advanced_calculator/advancedCalculator.cpp

```diff
@@ -0,0 +1,73 @@
+#include "advancedCalculator.hpp"
+
+#include <algorithm>
+#include <cctype>
+#include <cmath>
+#include <functional>
+#include <map>
+#include <regex>
+
+const std::map<std::string, std::function<double(double, double)>> commands{{"+", std::plus<double>()},
+                                                                            {"-", std::minus<double>()},
+                                                                            {"*", std::multiplies<double>()},
+                                                                            {"/", std::divides<double>()},
+                                                                            {"%", std::modulus<int>()},
+                                                                            {"^", [](double base, double exponent) { return pow(base, exponent); }},
+                                                                            {"$", [](double num, double root) { return pow(num, 1.0 / root); }}};
+
+void exportKeys(std::string& operators) {
+    for (const auto& el : commands) {
+        operators.append(el.first);
+    }
+}
+
+bool checkCharacters(const std::string& input) {
+    std::string allowedSigns{" ,.!"};
+    exportKeys(allowedSigns);
+
+    return std::any_of(input.cbegin(), input.cend(), [&allowedSigns](char letter) {
+        return std::find(allowedSigns.cbegin(), allowedSigns.cend(), letter) == allowedSigns.cend() && !isdigit(letter);
+    });
+}
+
+bool isNumFloat(const std::string& num) {
+    return stoi(num) != stod(num);
+}
+
+double factorial(double num) {
+    if (num <= 0.0) {
+        return 1.0;
+    }
+
+    return tgamma(num + 1.0);
+}
+
+ErrorCode process(std::string input, double* out) {
+    if (checkCharacters(input)) {
+        return ErrorCode::BadCharacter;
+    }
+
+    const std::regex commandRegex(R"((([-]?\d+[.]*\d*)[ ]*([+\-*\/$^%])[ ]*([-]?\d+[.]*\d*))|(([-]?\d+[.]*\d*)[ ]*!))");
```

Comment:
You can split this in 2 regex. One for factorial and other for rest cases. And this will cause regex will be more clearly to read. But This is only format suggestion :)